src/sources/internal_metrics.rs

```diff
@@ -0,0 +1,181 @@
+};
+use futures01::{sync::mpsc, Future, Sink};
+use metrics_core::Key;
+use metrics_runtime::{Controller, Measurement};
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, time::Duration};
+use stream_cancel::Tripwire;
+use tokio::timer::Interval;
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+pub struct InternalMetricsConfig;
+
+inventory::submit! {
+    SourceDescription::new::<InternalMetricsConfig>("internal_metrics")
+}
+
+#[typetag::serde(name = "internal_metrics")]
+impl SourceConfig for InternalMetricsConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let (trigger, tripwire) = Tripwire::new();
+        trigger.disable(); // TODO: don't actually run forever
+        let fut = run(get_controller()?, out, tripwire).boxed().compat();
+        Ok(Box::new(fut))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "internal_metrics"
+    }
+}
+
+fn get_controller() -> crate::Result<Controller> {
+    crate::metrics::CONTROLLER
+        .get()
+        .cloned()
+        .ok_or("metrics system not initialized".into())
+}
+
+async fn run(
+    controller: Controller,
+    mut out: mpsc::Sender<Event>,
+    mut tripwire: Tripwire,
+) -> Result<(), ()> {
+    let mut interval = Interval::new_interval(Duration::from_secs(2))
+        .compat()
+        .map(|_| ());
+
+    while let Some(()) = interval.next().await {
+        // Check for shutdown signal
+        if tripwire.poll().expect("polling tripwire").is_ready() {
+            break;
+        }
+
+        let metrics = capture_metrics(&controller);
+
+        let (sink, _) = out
+            .send_all(futures01::stream::iter_ok(metrics))
+            .compat()
+            .await
+            .expect("sending??");
+        out = sink;
+    }
+
+    Ok(())
+}
+
+fn capture_metrics(controller: &Controller) -> impl Iterator<Item = Event> {
+    controller
+        .snapshot()
+        .into_measurements()
+        .into_iter()
+        .map(|(k, m)| into_event(k, m))
+}
+
+fn into_event(key: Key, measurement: Measurement) -> Event {
+    let value = match measurement {
+        Measurement::Counter(v) => MetricValue::Counter { value: v as f64 },
+        Measurement::Gauge(v) => MetricValue::Gauge { value: v as f64 },
+        Measurement::Histogram(packed) => {
+            let values = packed
+                .decompress()
+                .into_iter()
+                .map(|i| i as f64)
+                .collect::<Vec<_>>();
+            let sample_rates = vec![1; values.len()];
+            MetricValue::Distribution {
+                values,
+                sample_rates,
+            }
+        }
+    };
+
+    let labels = key
+        .labels()
+        .map(|label| (String::from(label.key()), String::from(label.value())))
```

Comment:
this is unfortunate, I feel like labels don't really change that much would be nice if we could somehow do some sort of ref counting here or have some sort of pool of labels. Not a priority but just thinking.

Reply:
Agree, I can open another issue for this.